collector.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"io/ioutil"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/Clever/kayvee-go.v6/logger"
+
+	"github.com/Clever/ecs-task-metadata-exporter/data"
+	"github.com/Clever/ecs-task-metadata-exporter/metrics"
+)
+
+type collector struct {
+	Source data.Source
+	Logger logger.KayveeLogger
+}
+
+// NewCollector returns a prometheus.Collector configured to collect Docker metrics
+func NewCollector(source data.Source, l logger.KayveeLogger) prometheus.Collector {
+	// Set a logger with discarded output instead of nil, so we can call methods on log without panicing/checking for nil every time.
+	if l == nil {
+		l = logger.New("")
+		l.SetOutput(ioutil.Discard)
+	}
+	return collector{
+		Source: source,
+		Logger: l,
+	}
+}
+
+func (c collector) Describe(ch chan<- *prometheus.Desc) {
+	// By construction, no metrics will ever change, but at describe time, we haven't inspected the metadata yet, so we don't have the full list
+}
```

Comment:
Not understanding this. Is it supposed to be the following? Or are you saying that doesn't work because we have to do a different step first? This definitely doesn't seem to be `the super-set of all possible descriptors of metrics` https://pkg.go.dev/github.com/prometheus/client_golang/prometheus?tab=doc#Collector

```suggestion
func (c collector) Describe(ch chan<- *prometheus.Desc) {
	// By construction, no metrics will ever change, but at describe time, we haven't inspected the metadata yet, so we don't have the full list
	DescribeByCollect(c, ch)
}
```

Reply:
Ahh, I had some problems with using `DescribeByCollect` before, but I think it was because I was using https://pkg.go.dev/github.com/prometheus/client_golang/prometheus?tab=doc#NewPedanticRegistry instead of just `NewRegistry`? I'm not exactly sure. I sorta wonder if maybe something could happen if the sidecar container starts up and tries to initialize metrics before the main container exists or something...

Well it seems to work now, with `DescribeByCollect`, so I'll leave it and if there are any issues we can revisit.